src/odemis/driver/tfsbc.py

```diff
@@ -0,0 +1,357 @@
+
+    currLX = raw_to_current(rawLX)
+    currLY = raw_to_current(rawLY)
+    currUX = raw_to_current(rawUX)
+    currUY = raw_to_current(rawUY)
+
+    dc_xupper = currUX / (DCREVERSCOEF * DCROTUPPXX)
+    dc_xlower = currLX / (DCREVERSCOEF * DCROTLOWXX)
+    dc_yupper = currUY / (DCREVERSCOEF * DCROTUPPYY)
+    dc_ylower = currLY / (DCREVERSCOEF * DCROTLOWYY)
+
+    # Now we have to solve an overdetermined linear system of four equations with two variables.
+    A = numpy.array([[xupper[0], yupper[0]], [xlower[0], ylower[0]],
+                     [xupper[1], yupper[1]], [xlower[1], ylower[1]]])
+    b = numpy.array([dc_xupper, dc_xlower, dc_yupper, dc_ylower])
+    value, *_ = numpy.linalg.lstsq(A, b)
+
+    # A = numpy.array([[xupper[0], yupper[0]], [xlower[0], ylower[0]]])
+    # b = numpy.array([dc_xupper, dc_xlower])
+    # value = numpy.linalg.solve(A, b)
+    value = (value[0] * 1e-6, value[1] * 1e-6)  # µm --> m
+    return value
+
+
+class BeamShiftController(model.HwComponent):
+    """
+    Driver for the Thermofischer beam deflection controller.
+    This class provides the .shift VA containing a tuple of two floats which describe
+    the x and y beam offset in m in the stage coordinate system.
+
+    The conversion to internal ampere values (including scaling and rotation) is specified
+    through the MD_CALIB_BEAMSHIFT metadata (a 4x2 tuple, 4x (float, float)).
+    """
+
+    def __init__(self, name, role, port=None, serialnum=None, **kwargs):
+        """
+        :param port (str): port (e.g. /dev/ttyUSB0)
+        :param serialnum (str): serial number of RS485 adapter
+        """
+        # .hwVersion, .swVersion not available
+        model.HwComponent.__init__(self, name, role, **kwargs)
+
+        # Find port by RS485 adapter serial number
+        self._port = self._findDevice(port, serialnum)
+        self._serial = self._openSerialPort(self._port)
+
+        # Shift VA
+        # Range depends on metadata and will be checked in ._write_registers
+        self.shift = model.TupleContinuous((0, 0), range=((-1, -1), (1, 1)),
+                                           cls=(int, float), unit="m",
+                                           setter=self._setShift)
+
+    def _findDevice(self, ports=None, serialnum=None):
+        """
+        Look for a compatible device. Requires at least one of the arguments ports and serialnum.
+        ports (str): port (e.g. "/dev/ttyUSB0") or pattern for port ("/dev/ttyUSB*"), "/dev/fake" will start the simulator
+        serialnum (str): serial number
+        return (str): the name of the port used
+        raises:
+            ValueError: if no device on the ports with the given serial number is found
+        """
+        # At least one of the arguments ports and serialnum must be specified
+        if not ports and not serialnum:
+            raise ValueError("At least one of the arguments 'ports' and 'serialnum' must be specified.")
```

Comment:
The main class is going to raise this error if the user hasn't specified `port` or `serialnum`. As here it's spelled `ports`, this message is going to be confusing. I think that the easiest would be to just name the argument here also `port`.